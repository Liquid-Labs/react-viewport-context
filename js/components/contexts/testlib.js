import React from 'react'
import PropTypes from 'prop-types'

import { useViewportInfo } from '../hooks/useViewportInfo'

// following Material UI default theme 3.9.3
const defaultTheme = {
  breakpoints : {
    keys   : ['xs', 'sm', 'md', 'lg', 'xl'],
    values : {
      'xs' : 0,
      'sm' : 600,
      'md' : 960,
      'lg' : 1280,
      'xl' : 1920,
    }
  }
}

const weirdTheme = {
  breakpoints : {
    keys   : ['foo', 'bar', 'baz'],
    values : {
      'foo' : 0,
      'bar' : 82,
      'baz' : 105,
    }
  }
}

const ViewListener = ({callback}) => {
  const viewInfo = useViewportInfo()
  callback(viewInfo)

  return <div />
}

ViewListener.propTypes = {
  callback : PropTypes.func.isRequired
}

export { ViewListener, defaultTheme, weirdTheme }
