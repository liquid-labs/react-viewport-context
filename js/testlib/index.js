import React from 'react'
import PropTypes from 'prop-types'

import { useViewportInfo } from '../components/hooks/useViewportInfo'

// following Material UI default theme 3.9.3
const defaultTheme = {
  breakpoints : {
    keys   : ['xs', 'sm', 'md', 'lg', 'xl'],
    values : {
      'xs' : 0,
      'sm' : 600,
      'md' : 960,
      'lg' : 1280,
      'xl' : 1920,
    }
  },
  spacing : 8,
  typography : { useNextVariants : true },
  layout : {
    mainPadding : {
      'xs' : {
        top: 0,
        side: 0,
        bottom: 0,
      },
      'sm' : {
        top: 0.5,
        side: 0.5,
        bottom: 0.5,
      },
      'md' : {
        top: 0.5,
        side: 1,
        bottom: 1,
      },
      'lg' : {
        top: 1,
        side: 1,
        bottom: 1,
      },
      'xl' : {
        top: 1,
        side: 1,
        bottom: 1,
      },
    }
  },
}

const weirdTheme = {
  breakpoints : {
    keys   : ['foo', 'bar', 'baz'],
    values : {
      'foo' : 0,
      'bar' : 82,
      'baz' : 105,
    }
  },
  spacing : 5,
  typography : { useNextVariants : true },
  layout : {
    mainPadding : {
      'xs' : {
        top: 0.3,
        side: 0.1,
        bottom: 4,
      },
      'sm' : {
        top: 0.5,
        side: 0.25,
        bottom: 0.5,
      },
      'md' : {
        top: 0.5,
        side: 1,
        bottom: 1,
      },
      'lg' : {
        top: 1,
        side: 1,
        bottom: 1,
      },
      'xl' : {
        top: 1,
        side: 1,
        bottom: 1,
      },
    }
  },
}

const ViewListener = ({callback}) => {
  const viewInfo = useViewportInfo()
  callback(viewInfo)

  return <div />
}

ViewListener.propTypes = {
  callback : PropTypes.func.isRequired
}

export { ViewListener, defaultTheme, weirdTheme }
