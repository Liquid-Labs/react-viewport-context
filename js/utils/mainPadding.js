const paddingSpec = (spacingUnit, relSpec) => ({
  top    : Math.ceil(relSpec.top * spacingUnit),
  side   : Math.ceil(relSpec.side * spacingUnit),
  bottom : Math.ceil(relSpec.bottom * spacingUnit),
})

const getMainPaddingSpec = (theme) => {
  const breakpointKeys = theme.breakpoints.keys
  const spacingUnit = theme.spacing.unit // a number in pixels
  const themeSpec = theme.layout.mainPadding // specifies relative to unit for each breakpoint

  return breakpointKeys.reduce((acc, key) =>
      (acc[key] = paddingSpec(spacingUnit, themeSpec[key]))
        && acc,
    {}
  )
}

const mainPaddingStyles = (theme) => {
  const mainPaddingSpec = getMainPaddingSpec(theme)
  const breakpointKeys = theme.breakpoints.keys
/*
  console.log(breakpointKeys.reduce((acc, key) =>
    (acc[theme.breakpoints.up(key)] = {
      paddingTop  : `${mainPaddingSpec[key]['top']}px`,
    }) && acc,
  {}))

  return {
    mainPaddingTop : breakpointKeys.reduce((acc, key) =>
      (acc[theme.breakpoints.up(key)] = {
        paddingTop  : `${mainPaddingSpec[key]['top']}px`,
      }) && acc,
    {}),
    mainPaddingBottom : {
      paddingBottom: '5px',
    },
    mainPaddingSides : {
      paddingLeft : '5px',
      paddingRight : '5px'
    }
  }
*/
  return {
    mainPaddingSides : breakpointKeys.reduce((acc, key) =>
      (acc[theme.breakpoints.up(key)] = {
        paddingLeft  : `${mainPaddingSpec[key]['side']}px`,
        paddingRight : `${mainPaddingSpec[key]['side']}px`,
      }) && acc,
    {}),
    mainPaddingTop : breakpointKeys.reduce((acc, key) =>
      (acc[theme.breakpoints.up(key)] = {
        paddingTop  : `${mainPaddingSpec[key]['top']}px`,
      }) && acc,
    {}),
    mainPaddingBottom : breakpointKeys.reduce((acc, key) =>
      (acc[theme.breakpoints.up(key)] = {
        paddingBottom  : `${mainPaddingSpec[key]['bottom']}px`,
      }) && acc,
    {})
  }
}

export { getMainPaddingSpec, mainPaddingStyles }
