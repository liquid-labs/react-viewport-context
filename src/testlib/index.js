import React from 'react' /* eslint-disable-line no-unused-vars */
import PropTypes from 'prop-types'

import { useWindowInfo } from '../components/hooks/use-window-info'

// following Material UI default theme 3.9.3
const defaultTheme = {
  breakpoints : {
    values : {
      xs : 0,
      sm : 600,
      md : 960,
      lg : 1280,
      xl : 1920
    }
  },
  spacing    : 8,
  typography : { useNextVariants : true },
  layout     : {
    mainPadding : {
      xs : {
        top    : 0,
        side   : 0,
        bottom : 0
      },
      sm : {
        top    : 0.5,
        side   : 0.5,
        bottom : 0.5
      },
      md : {
        top    : 0.5,
        side   : 1,
        bottom : 1
      },
      lg : {
        top    : 1,
        side   : 1,
        bottom : 1
      },
      xl : {
        top    : 1,
        side   : 1,
        bottom : 1
      }
    }
  }
}

const weirdTheme = {
  breakpoints : {
    values : {
      foo : 0,
      bar : 82,
      baz : 105
    }
  },
  spacing    : 5,
  typography : { useNextVariants : true },
  layout     : {
    mainPadding : {
      xs : {
        top    : 0.3,
        side   : 0.1,
        bottom : 4
      },
      sm : {
        top    : 0.5,
        side   : 0.25,
        bottom : 0.5
      },
      md : {
        top    : 0.5,
        side   : 1,
        bottom : 1
      },
      lg : {
        top    : 1,
        side   : 1,
        bottom : 1
      },
      xl : {
        top    : 1,
        side   : 1,
        bottom : 1
      }
    }
  }
}

const ViewListener = ({ callback }) => {
  const viewInfo = useWindowInfo()

  callback(viewInfo)

  return <div />
}

ViewListener.propTypes = {
  callback : PropTypes.func.isRequired
}

export { ViewListener, defaultTheme, weirdTheme }
