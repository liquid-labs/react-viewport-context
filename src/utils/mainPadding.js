/**
 * @file Here we define helper functions to generate the proper style based on 'theme.layout.mainPadding' breakpoint
 * specific spacing factors.
 */

/**
 * Determines the actual pixel paddings given <code>spacing</code> and <code>relSpec</code>.
 * @param {string | Function} spacing - Either a number to be multiplied by each <code>resSpec</code> weighting or a
 *   function which does the same thing
 * @param {object} relSpec - An object representing the relative space (determined by <code>spacing</code>) to apply to
 *   the top, side, and bottom.
 * @param {number} relSpec.bottom - Bottom weighting.
 * @param {number} relSpec.side - Side weighting.
 * @param {number} relSpec.top - Top weighting.
 * @returns {object} An object with actual 'px' settings for <code>top</code>, <code>side</code>, and <code>
 *   bottom</code>.
 * 
 * @memberof react-viewport-context
 */
const paddingSpec = (spacing, relSpec) => {
  return typeof spacing === 'number'
    ? {
        top    : Math.ceil(spacing * relSpec.top) + 'px',
        side   : Math.ceil(spacing * relSpec.side) + 'px',
        bottom : Math.ceil(spacing * relSpec.bottom) + 'px'
      }
    : {
        top    : spacing(relSpec.top),
        side   : spacing(relSpec.side),
        bottom : spacing(relSpec.bottom)
      }
}

const getMainPaddingSpec = (theme) => {
  const breakpointKeys = Object.keys(theme.breakpoints.values)
  const themeSpec = theme.layout.mainPadding // specifies relative to unit for each breakpoint

  return breakpointKeys.reduce((acc, key) =>
    (acc[key] = paddingSpec(theme.spacing, themeSpec[key])) &&
        acc,
  {}
  )
}

const mainPaddingStyles = (theme, classes) => {
  const mainPaddingSpec = getMainPaddingSpec(theme)
  const breakpointKeys = Object.keys(theme.breakpoints.values)
  return {
    [classes.mainPaddingSides] : breakpointKeys.reduce((acc, key) =>
      (acc[theme.breakpoints.up(key)] = {
        paddingLeft  : mainPaddingSpec[key].side,
        paddingRight : mainPaddingSpec[key].side
      }) && acc,
    {}),
    [classes.mainPaddingTop] : breakpointKeys.reduce((acc, key) =>
      (acc[theme.breakpoints.up(key)] = {
        paddingTop : mainPaddingSpec[key].top
      }) && acc,
    {}),
    [classes.mainPaddingBottom] : breakpointKeys.reduce((acc, key) =>
      (acc[theme.breakpoints.up(key)] = {
        paddingBottom : mainPaddingSpec[key].bottom
      }) && acc,
    {})
  }
}

export { getMainPaddingSpec, mainPaddingStyles }
