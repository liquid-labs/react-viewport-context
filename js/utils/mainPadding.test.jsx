/* global afterEach describe Event expect test */
import React from 'react'
import classNames from 'classnames'

import { mainPaddingStyles } from './mainPadding'

import { cleanup, render } from 'react-testing-library'
import { defaultTheme } from '../testlib'

import { SheetsRegistry } from 'jss'

import { createTheme } from '@mui/material/styles'
import { makeStyles, StylesProvider, ThemeProvider } from '@mui/styles'

const useMainPaddingStyles = makeStyles(mainPaddingStyles)

const TestMain = (props) => {
  const classes = useMainPaddingStyles()
  const className = classNames(
    classes.mainPaddingTop,
    classes.mainPaddingSides,
    classes.mainPaddingBottom)

  return <div className={className} {...props}>Hi!</div>
}

const defaultMuiTheme = createTheme(defaultTheme)

describe("mainPaddingStyles", () => {
  afterEach(cleanup)

  test("generates expected padding styles", () => {
    const sheets = new SheetsRegistry()

    const { getByTestId } = render(
      <StylesProvider sheetsRegistry={sheets}>
        <ThemeProvider theme={defaultMuiTheme}>
            <TestMain data-testid='main'/>
        </ThemeProvider>
      </StylesProvider>
    )

    expect(sheets.toString()).toMatchSnapshot()
  })
})
