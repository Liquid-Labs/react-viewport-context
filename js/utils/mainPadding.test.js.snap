// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mainPaddingStyles generates expected padding styles 1`] = `
"@media (min-width:0px) {
  .Hook-mainPaddingSides-8huz73 {
    padding-left: 0px;
    padding-right: 0px;
  }
}
@media (min-width:600px) {
  .Hook-mainPaddingSides-8huz73 {
    padding-left: 4px;
    padding-right: 4px;
  }
}
@media (min-width:960px) {
  .Hook-mainPaddingSides-8huz73 {
    padding-left: 8px;
    padding-right: 8px;
  }
}
@media (min-width:1280px) {
  .Hook-mainPaddingSides-8huz73 {
    padding-left: 8px;
    padding-right: 8px;
  }
}
@media (min-width:1920px) {
  .Hook-mainPaddingSides-8huz73 {
    padding-left: 8px;
    padding-right: 8px;
  }
}
@media (min-width:0px) {
  .Hook-mainPaddingTop-vwmiep {
    padding-top: 0px;
  }
}
@media (min-width:600px) {
  .Hook-mainPaddingTop-vwmiep {
    padding-top: 4px;
  }
}
@media (min-width:960px) {
  .Hook-mainPaddingTop-vwmiep {
    padding-top: 4px;
  }
}
@media (min-width:1280px) {
  .Hook-mainPaddingTop-vwmiep {
    padding-top: 8px;
  }
}
@media (min-width:1920px) {
  .Hook-mainPaddingTop-vwmiep {
    padding-top: 8px;
  }
}
@media (min-width:0px) {
  .Hook-mainPaddingBottom-1a6gl0k {
    padding-bottom: 0px;
  }
}
@media (min-width:600px) {
  .Hook-mainPaddingBottom-1a6gl0k {
    padding-bottom: 4px;
  }
}
@media (min-width:960px) {
  .Hook-mainPaddingBottom-1a6gl0k {
    padding-bottom: 8px;
  }
}
@media (min-width:1280px) {
  .Hook-mainPaddingBottom-1a6gl0k {
    padding-bottom: 8px;
  }
}
@media (min-width:1920px) {
  .Hook-mainPaddingBottom-1a6gl0k {
    padding-bottom: 8px;
  }
}"
`;
